File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:        
        def dfs(node, left, right):
            nonlocal ans
            if not node:
                return
            ans = max(ans, left, right)
            if node.left:
                dfs(node.left, right + 1, 0)
            if node.right:
                dfs(node.right, 0, left + 1)
                
        
        ans = 0
        dfs(root, 0, 0)
        
        return ans

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestZigZag = function(root) {
  let ans = 0;
  
  const dfs = (node, left, right) => {
    if (!node) return;
    
    ans = Math.max(ans, left, right);
    if (node.left) dfs(node.left, right + 1, 0);
    if (node.right) dfs(node.right, 0, left + 1);
  }
  
  dfs(root, 0, 0);
  
  return ans;
};